### PR TITLE
[BEAM-3326] Add `SdkHarnessClientControlService`

### PR DESCRIPTION
**Pull request by tgroh**
This is the higher-level Control Service implementation which vends
SdkHarnessClient instances.

Implement AutoCloseable on SdkHarnessClient to close the underlying
control client when the high-level service is shut down.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand:
   - [x] What the pull request does
   - [x] Why it does it
   - [x] How it does it
   - [x] Why this approach
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClientControlService.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted while waiting for client", e);
+    }
+  }
+
+  @Override
+  public void close() throws Exception {
+    for (SdkHarnessClient client : activeClients) {
+      client.close();
```
Comment (kennknowles) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClientControlService.java]: Can you loudly document on the `FnApiControlClientPoolService` that whoever grabs a handle to a client owns closing it?

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClient.java [hunk 2]
```diff
@@ -50,7 +50,7 @@
  * <p>This provides a Java-friendly wrapper around {@link FnApiControlClient} and {@link
  * CloseableFnDataReceiver}, which handle lower-level gRPC message wrangling.
  */
-public class SdkHarnessClient {
+public class SdkHarnessClient implements AutoCloseable {
```
Comment (kennknowles) [inline runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClient.java]: Just for my edification: if I am authoring a runner and I currently interact directly with a `FnApiControlClient` that I got from a `FnApiControlClientPoolService`, should I instead be trying to use an `SdkHarnessClient` that I got from an `SdkHarnessClientPoolService`?

Comment (tgroh) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClient.java]: Yes. `FnApiControlClientPoolService` shouldn't really be visible outside of the `control` package

File: runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClientControlService.java (hunk 1 above)
Comment (tgroh) [reply runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/SdkHarnessClientControlService.java]: Done.

**Comment by kennknowles** [comment]
Filed tickets for the flakes.

**Comment by kennknowles** [comment]
run java gradle precommit

**Pull request merged**